File: Fixing Two swapped nodes of a BST.java
class Solution {
    ArrayList<Integer> list;
    int i;
    
    public void inorder(Node node){
        if(node==null){
            return;
        }
        
        //L
        inorder(node.left);
        
        //D
        list.add(node.data);
        
        //R
        inorder(node.right);
    }
    
    public void solve(Node node){
        if(node==null){
            return;
        }
        
        //L
        solve(node.left);
        
        //D
        int ele=list.get(i);
        i++;
        node.data=ele;
        
        //R
        solve(node.right);
    }
    
    public Node correctBST(Node root) {
        list=new ArrayList<>();
        i=0;
        
        //inorder
        inorder(root);
        
        //sort
        Collections.sort(list);
        
        //solve
        solve(root);
        
        return root;
    }
}

File: Fixing Two swapped nodes of a BST.cpp
class Solution {
public:
    struct Node* first = NULL;
    struct Node* second = NULL;
    struct Node* prev = NULL;
    
    void inOrder(struct Node* root){
        if(root == NULL){
            return;
        }
        inOrder(root->left);
        if(prev != NULL && root->data < prev->data){
            if(first == NULL){
                first = prev;
            }
            second = root;
        }
        prev = root;
        inOrder(root->right);
    }
    
    struct Node* correctBST(struct Node* root) {
        inOrder(root);
        int temp = first->data;
        first->data = second->data;
        second->data = temp;
        return root;
    }
};
